internal/engine/planner/planner.go

```diff
@@ -236,20 +243,25 @@ func (rpe *ResourcePolicyEvaluator) EvaluateResourcesQueryPlan(ctx context.Conte
 
 			derivedRoles = append(derivedRoles, rN{
 				role: drName,
-				f: func() (*qpN, error) {
+				node: memoize(func() (*qpN, error) {
 					if dr.Condition == nil {
 						return mkTrueNode(), nil
 					}
-					node, err := evaluateCondition(dr.Condition, input, rpe.Globals, drVariables)
+					node, err := evaluateCondition(dr.Condition, request, rpe.Globals, drVariables, derivedRolesList)
 					if err != nil {
 						return nil, err
 					}
 					return node, nil
-				},
-				node: nil,
+				}),
 			})
 		}
 
+		sort.Slice(derivedRoles, func(i, j int) bool {
+			return derivedRoles[i].role < derivedRoles[j].role
+		})
```

Comment:
Again, probably not necessary?

Reply:
In this case, I actually think this is beneficial outside of the `P.derivedRoles` implementation. By having a nondeterministic ordering of derived roles, the output from the planner is not stable (given a rule that references multiple derived roles, the order that the corresponding conditions appear in the AST is similarly nondeterministic).

Having a stable AST is potentially desirable for consumers who want to be able to transform the query plan to some executable form and cache the result?

Given that there are only going to be a handful of roles in the list, I don't think the overhead is too bad.